# Problemas de prueba de hipótesis

## Ejercicio 1

Usted es un nutricionista que investiga dos tipos diferentes de dietas para ver si existe una diferencia significativa en la pérdida de peso después de un mes. Eliges dos grupos aleatorios de personas; un grupo sigue la primera dieta y el otro sigue la segunda. Al final del mes, se registra la pérdida de peso (en kg) de cada persona:

| Diet 1 | Diet 2 |
|:-------|:-------|
| 2.0 | 3.0 |
| 2.5 | 3.2 |
| 3.0 | 3.1 |
| 2.8 | 2.9 |
| 2.3 | 2.8 |
| 2.7 | 3.0 |
| 2.5 | 3.2 |

Con estos datos se busca responder a la siguiente pregunta: ¿Existe una diferencia significativa en la pérdida de peso promedio entre las personas que siguieron la primera dieta y las que siguieron la segunda dieta?

Para sacar conclusiones, siga los siguientes puntos:

-Enunciar la hipótesis: hipótesis nula y alternativa.
-Realizar la prueba para comprobar la hipótesis. Puede utilizar la prueba t de Student.
-Analizar las conclusiones.


In [68]:
from scipy import stats
import numpy as np
import statsmodels.api as sm
import statsmodels.stats.multicomp
from scipy.stats import f_oneway


In [ ]:
# TODO

diet_1 = [2.0, 2.5, 3.0, 2.8, 2.3, 2.7, 2.5]
diet_2 = [3.0, 3.2, 3.1, 2.9, 2.8, 3.0, 3.2]

H_0: No hay una diferencia significativa en la pérdida de peso promedio entre las personas que siguieron la primera dieta y las que siguieron la segunda dieta

H_1: Hay una diferencia significativa en la pérdida de peso promedio entre las personas que siguieron la primera dieta y las que siguieron la segunda dieta.

In [70]:
#

In [71]:

# Comprobamos con T test : 

pval = stats.ttest_ind(diet_1, diet_2).pvalue
alpha = 0.05

if pval < alpha:
    print("La evidencia de los datos es suficientemente fuerte como para rechazar la hipótesis nula H₀, es decir, respaldan la Hipótesis alternativa (H₁),\n que no hay significativa en la pérdida de peso promedio entre las personas que siguieron la primera dieta y las que siguieron la segunda dieta")
    print(f"Es muy improbable que la evidencia observada se deba al azar (p-valor={pval}).")
else:
    print("No hay suficiente evidencia para rechazar la hipótesis nula H₀, es decir, no hay suficiente evidencia para afirmar que la hipótesis nula es falsa.\nAsumimos entonces que hay posibilidad todavía que haya diferencia significativa en la perdida de peso promedio entre las personas que siguieron la 1er dieta y las que siguoeron la 2nda dieta.")
    print("Esto no implica que aceptemos H₀, ya que la ausencia de evidencia no es lo mismo que evidencia en su favor.")
    print(f"Es muy probable que la evidencia observada se deba al azar (p-valor={pval}).")



La evidencia de los datos es suficientemente fuerte como para rechazar la hipótesis nula H₀, es decir, respaldan la Hipótesis alternativa (H₁),
 que no hay significativa en la pérdida de peso promedio entre las personas que siguieron la primera dieta y las que siguieron la segunda dieta
Es muy improbable que la evidencia observada se deba al azar (p-valor=0.004083270191713912).


## ANOVA

**ANOVA** (*Analysis of Variance - Análisis de variación*) Es una técnica estadística utilizada para comparar las medidas de dos o más grupos. La idea detrás de ANOVA es descomponer la variabilidad total de los datos en dos componentes: variabilidad entre grupos y variabilidad dentro de los grupos:

- **Variabilidad entre grupos**: Esta variabilidad se refiere a las diferencias entre las medias del grupo. Si esta variabilidad es considerablemente mayor que la variabilidad dentro del grupo, podría ser una indicación de que al menos una de las medias del grupo es diferente.
- **Variabilidad dentro del grupo**: Esta variabilidad se refiere a la dispersión de los datos dentro de cada grupo. Si todos los grupos tienen una variabilidad similar, entonces cualquier diferencia notable en las medias de los grupos podría considerarse significativa.

Las hipótesis en ANOVA suelen incluir:

- **Hipótesis nula** ($H₀$): Las medias de todos los grupos son iguales.
- **Hipótesis alternativa** ($H₁$): Al menos una de las medias del grupo es diferente.

Si el resultado de la prueba ANOVA es significativo (por ejemplo, un valor p inferior a un umbral como 0,05), esto sugiere que al menos la media de un grupo es diferente.

## Ejercicio 2

Un agricultor decide probar tres tipos diferentes de fertilizantes para determinar si alguno es superior en términos de producción de maíz. El agricultor planta maíz en 15 parcelas idénticas y utiliza los tres fertilizantes (5 parcelas para cada tipo). Al final de la temporada mide el rendimiento de maíz (en kg) de cada parcela, con el siguiente resultado:

| Fertilizante 1 | Fertilizante 2 | Fertilizante 3 |
|:-------------|:-------------|:-------------|
| 20 | 22 | 24 |
| 21 | 21 | 23 |
| 20 | 23 | 22 |
| 19 | 22 | 23 |
| 20 | 21 | 24 |

Con estos datos busca responder la siguiente pregunta: ¿Existe una diferencia significativa en el rendimiento promedio del maíz entre los tres tipos de fertilizantes?

Para ayudarte, sigue los siguientes puntos:

-Enunciar la hipótesis: hipótesis nula y alternativa.
-Realizar la prueba ANOVA.
-Analizar las conclusiones.
-Si un fertilizante es mejor que otro, ¿cómo podemos saberlo?

In [72]:
# TODO

#H_0: Las medias de todos los grupos son iguales y un fertilizante superior

#H_1: Al menos una de las medias del grupo es diferente.

fertilizante_1 = [20, 21, 20, 19, 20]
fertilizante_2 = [22, 21, 23, 22, 21]
fertilizante_3 = [24, 23, 22, 23, 24]


pvalor  = stats.f_oneway(fertilizante_1, fertilizante_2, fertilizante_3).pvalue
alpha = 0.05

if pvalor < alpha:
    print(f"Rechazamos H₀: Hay una diferencia significativa entre los fertilizantes. (valor p = {pvalor})")
else:
    print(f"No rechazamos H₀: No hay diferencia significativa entre los fertilizantes. (valor p = {pvalor})")




Rechazamos H₀: Hay una diferencia significativa entre los fertilizantes. (valor p = 0.00014047824793190475)


In [73]:
#otra manera :

F, p = stats.f_oneway(fertilizante_1, fertilizante_2, fertilizante_3)
print(f"Variabilidad entre grupos / variabilidad dentro de los grupos: {F}")
print(f"Pvalor: {p}")

if p< alpha:
    print(f"Rechazamos H₀: Hay una diferencia significativa entre los fertilizantes. (valor p = {p})")
else:
    print(f"No rechazamos H₀: No hay diferencia significativa entre los fertilizantes. (valor p = {p})")

Variabilidad entre grupos / variabilidad dentro de los grupos: 20.315789473684188
Pvalor: 0.00014047824793190475
Rechazamos H₀: Hay una diferencia significativa entre los fertilizantes. (valor p = 0.00014047824793190475)


In [74]:
# Usamos Tukey : Calculate all pairwise comparisons with TukeyHSD confidence intervals
# compara cada par de grupos y te dice cuáles son significativamente diferentes entre sí.
# statsmodels.stats.multicomp.pairwise_tukeyhsd(endog, groups, alpha=0.05)

datos = np.array(fertilizante_1 + fertilizante_2 + fertilizante_3)
grupos = ['Fertilizante 1'] * 5 + ['Fertilizante 2'] * 5 + ['Fertilizante 3'] * 5

tukey_resultados = statsmodels.stats.multicomp.pairwise_tukeyhsd(datos, grupos, alpha=0.05)
print(tukey_resultados)

       Multiple Comparison of Means - Tukey HSD, FWER=0.05        
    group1         group2     meandiff p-adj  lower  upper  reject
------------------------------------------------------------------
Fertilizante 1 Fertilizante 2      1.8 0.0099 0.4572 3.1428   True
Fertilizante 1 Fertilizante 3      3.2 0.0001 1.8572 4.5428   True
Fertilizante 2 Fertilizante 3      1.4 0.0409 0.0572 2.7428   True
------------------------------------------------------------------


**meandiff** : diferencia de medias 

**p-adj** = --> p, es lo que se compara con alpha, si es menor entonces indica que la diferencia es significativa

**Lower y Upper**: Definen el rango del intervalo de confianza para la diferencia de medias. Si este intervalo no incluye cero, indica que la diferencia es significativa.

**Reject**: Si es True, significa que la diferencia entre los grupos comparados es significativa (es decir, se rechaza la hipótesis nula de que las medias son iguales). 
Si es False, no hay evidencia suficiente para rechazar la hipótesis nula y no se puede concluir que haya una diferencia significativa.

**Entonces en ese caso**

Todas las comparaciones tienen "True" en Reject --> todas las diferencias entre los grupos comparados son significativas
Ningun intervalo de confianza incluye 0 --> todas las diferencias entre los grupos comparados son significativas
= Los tres fertilizantes muestran diferencias significativas en cuanto a rendimiento. 

Ahora para saber cual es más eficaz miramos meandiff y p : 

El fertilizante 3 tiene una diferencia significativa de 3.2 con el fertilizante 1 (p = 0.0001) y una diferencia significativa de 1.4 con el fertilizante 2 (p= 0.0409).
En los 2 casos, el fertilizante 3 tiene un rendimiento superior, entonces se puede considerar que es el mejor.
